src/sync.rs

```diff
@@ -208,20 +208,28 @@ trait EvalContextExtPriv<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 // situations.
 impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
-    #[inline]
-    /// Peek the id of the next mutex
-    fn mutex_next_id(&self) -> MutexId {
-        let this = self.eval_context_ref();
-        this.machine.threads.sync.mutexes.next_index()
-    }
-
     #[inline]
     /// Create state for a new mutex.
     fn mutex_create(&mut self) -> MutexId {
         let this = self.eval_context_mut();
         this.machine.threads.sync.mutexes.push(Default::default())
     }
 
+    #[inline]
+    /// Provides the closure with the next MutexId. Creates that mutex if the closure returns None,
+    /// otherwise returns the value from the closure
+    fn mutex_get_or_create<F>(&mut self, existing: F) -> InterpResult<'tcx, MutexId>
+    where
+        F: FnOnce(&mut MiriEvalContext<'mir, 'tcx>, MutexId) -> InterpResult<'tcx, Option<MutexId>>,
+    {
+        let this = self.eval_context_mut();
+        if let Some(old) = existing(this, this.machine.threads.sync.mutexes.next_index())? {
+            Ok(old)
+        } else {
+            Ok(self.mutex_create())
```

Comment:
I like the closure-based API, but this still spreads the contract over multiple functions.

Can you inline `mutex_create` here? Then at least the two accesses to `this.machine.threads.sync.mutexes` (`next_index` and `push`) are in the same function.

We *do* give control to the user in between via a closure, which has access to `this`, so there is still no full guarantee that we return the right ID here... if `existing` calls back into `mutex_get_or_create`, the return value is wrong. So I'd prefer to still also see an assertion.